api/bibliotheca.py

```diff
@@ -1413,96 +1461,218 @@ def timestamp(self):
+        :type achievements: str
+        """
+        progress.start = start
+        progress.finish = finish
+        progress.achievements = achievements
+        progress.finalize(
+            service=self.service_name,
+            service_type=Timestamp.MONITOR_TYPE,
+            collection=self.collection,
+        )
+        progress.apply(self._db)
+        self._db.commit()
+
+    def purchases(self, start, end):
+        """Ask Bibliotheca for a MARC record for each book purchased
+        between `start` and `end`.
+
+        :yield: A sequence of pymarc Record objects
+        """
+        offset = 1     # Smallest allowed offset
+        page_size = 50 # Maximum supported size.
+        records = None
+        while records is None or len(records) >= page_size:
+            records = [
+                x for x in self.api.marc_request(start, end, offset, page_size)
+            ]
+            for record in records:
+                yield record
+            offset += page_size
+
+    def process_record(self, record, purchase_time):
+        """Record the purchase of a new title.
+
+        :param record: Bibliographic information about the new title.
+        :type record: pymarc.Record
+
+        :param purchase_time: Put down this time as the time the 
+           purchase happened.
+        :type start_time: datetime.datetime
+
+        :return: A LicensePool representing the new title.
+        :rtype: core.model.LicensePool
+        """
+        # The control number associated with the MARC record is what
+        # we call the Bibliotheca ID.
+        control_numbers = [x for x in record.fields if x.tag == '001']
+        # These errors should not happen in real usage.
+        error = None
+        if not control_numbers:
+            error = "Ignoring MARC record with no Bibliotheca control number."
+        elif len(control_numbers) > 1:
+            error = "Ignoring MARC record with multiple Bibliotheca control numbers."
+        if error is not None:
+            self.log.error(error + " " + record.as_json())
             return
-        raise ValueError('A timespan greater than 72 hours can result in spurious'
-                         ' results from the Bibliotheca events API.')
 
-    def catch_up_from(self, start, cutoff, progress):
-        # This method operates in three modes: Uninitialized, Backlog,
-        # and Recent. Uninitialized Mode happens when the monitor's
-        # `progress.counter` is null, but is functionally identical
-        # to Backlog Mode (0, backlog_mode_counter_value).
-        # In both of these modes, the timespan for which we fetch
-        # events is `backlog_mode_timespan` and it is NOT considered
-        # an error when the response contains no events.
-        # In Recent Mode, the timespan (`backlog_mode_timespan`) is
-        # longer and a response contains no events DOES raise an error.
-        recent_mode_counter_value = 1
-        backlog_mode_counter_value = 0
```

Comment:
I was able to remove the code that switched from one mode to another, since it's not an issue for the BibliothecaPurchaseMonitor, and the BibliothecaEventMonitor no longer needs to be super reliable.